### PR TITLE
[in_app_purchase] Add alternative billing apis for android

### PR DESCRIPTION
**Pull request by reidbaker**
- Update the emulator versions and expose cipd. (#6025
- Enable alternitive billing only available check, add test and code to handle service unavilable in getBillingConfig
- Enable alternative billing only during client creation and tests covering fallback path
- ShowAlternativeBillingDialog android native method added
- Add tests for null activity behavior
- Remove not needed lines of code
- Add showAlternativeBillingOnlyInformationDialog and isAlternativeBillingOnlyAvailable to android platform addition and billing client wrapper.
- test showAlternativeBillingOnlyInformationDialog and isAlternativeBillingOnlyAvailable in platfrom addition and billing_client

Fixes flutter/flutter/issues/142618

Still left TODO: 
* [x] incorporate new apis into example app 
* [x] expose alternative billing only [dart api](https://github.com/flutter/packages/pull/6056/files/d4c445422f2cd3f0627f575d85b59b559e0e9f69#r1480455450)
* [x] Expose alternative billing reporting details
* [ ] Configure end to end working example with playstore

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_android/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.3.0+19
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/CHANGELOG.md:1]: Nit: As a new feature it should be incrementing the "minor" version (which for <1.0 in Dart convention is the third instead of the second), so this should be `0.3.1`.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart [hunk 2]
```diff
@@ -157,4 +157,26 @@ class InAppPurchaseAndroidPlatformAddition
         .runWithClient((BillingClient client) => client.getBillingConfig());
     return billingConfig.countryCode;
   }
+
+  /// Returns if the caller can use alternative billing only without giving the
+  /// user a choice to use Play billing.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart:162]: Do we expect to also need to support the with-user-choice version, or are we waiting to see if it's frequently requested first? I'm wondering if we should have an issue reference or anything here.

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java [hunk 3]
```diff
@@ -65,10 +66,27 @@ static final class MethodNames {
     static final String IS_FEATURE_SUPPORTED = "BillingClient#isFeatureSupported(String)";
     static final String GET_CONNECTION_STATE = "BillingClient#getConnectionState()";
     static final String GET_BILLING_CONFIG = "BillingClient#getBillingConfig()";
+    static final String IS_ALTERNATIVE_BILLING_ONLY_AVAILABLE =
+        "BillingClient#isAlternativeBillingOnlyAvailable()";
+    static final String CREATE_ALTERNATIVE_BILLING_ONLY_REPORTING_DETAILS =
+        "BillingClient#createAlternativeBillingOnlyReportingDetails()";
+    static final String SHOW_ALTERNATIVE_BILLING_ONLY_INFORMATION_DIALOG =
+        "BillingClient#showAlternativeBillingOnlyInformationDialog()";
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java:74]: For the record since I've missed this before: according to find in page, these three strings exactly match the Dart versions.

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java [hunk 4]
```diff
@@ -147,7 +166,13 @@ public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result
         isReady(result);
         break;
       case MethodNames.START_CONNECTION:
-        startConnection((int) call.argument("handle"), result);
+        final int handle = (int) call.argument(MethodArgs.HANDLE);
+        boolean enableAlternativeBillingOnly = false;
+        if (call.hasArgument(MethodArgs.ENABLE_ALTERNATIVE_BILLING_ONLY)
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java:171]: I would expect to see changes to https://github.com/flutter/packages/blob/d4c445422f2cd3f0627f575d85b59b559e0e9f69/packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart#L119-L123 for this to be able to do anything. Am I looking at the wrong part of the Dart code? If not, this will never be true.

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java [hunk 5]
```diff
@@ -240,6 +241,16 @@ static HashMap<String, Object> fromBillingConfig(
     return info;
   }
 
+  /**
+   * Converter from {@link BillingResult} and {@link AlternativeBillingOnlyReportingDetails} to map.
+   */
+  static HashMap<String, Object> fromAlternativeBillingOnlyReportingDetails(
+      BillingResult result, AlternativeBillingOnlyReportingDetails details) {
+    HashMap<String, Object> info = fromBillingResult(result);
+    info.put("externalTransactionToken", details.getExternalTransactionToken());
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java:250]: This just has to match the field name on the Dart side due to JSON generation, right? (I.e., it's expected that I don't see a matching string constant in Dart?)

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/alternative_billing_only_reporting_details_wrapper.dart [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import '../../billing_client_wrappers.dart';
+
+// WARNING: Changes to `@JsonSerializable` classes need to be reflected in the
+// below generated file. Run `flutter packages pub run build_runner watch` to
+// rebuild and watch for further changes.
+part 'alternative_billing_only_reporting_details_wrapper.g.dart';
+
+/// The error message shown when the map representing details is invalid from method channel.
+///
+/// This usually indicates a serious underlining code issue in the plugin.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/alternative_billing_only_reporting_details_wrapper.dart:17]: underlying?

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 7]
```diff
@@ -340,6 +341,57 @@ class BillingClient {
...
 
+  /// isAlternativeBillingOnlyAvailable method channel string identifier.
+  //
+  // Must match the value of IS_ALTERNATIVE_BILLING_ONLY_AVAILABLE in
+  // ../../../android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java
+  @visibleForTesting
+  static const String isAlternativeBillingOnlyAvailableMethodString =
+      'BillingClient#isAlternativeBillingOnlyAvailable()';
+
+  /// Checks if "AlterntitiveBillingOnly" features is available.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart:352]: Typo: "feature is"

(or "features are")

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart:1]: There should be new testing in the `startConnection` group if it's supposed to be sending new parameters.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java (hunk 3 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java:74]: Thanks for the double check!

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart (hunk 2 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart:162]: Yes that is the next pr :) 
https://github.com/flutter/flutter/issues/143004

My intention is to expose all the newly added functionality in 6.1 this month.

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java (hunk 4 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java:171]: You are correct, this slipped by me between last friday and yesterday. I would have caught it when adding functionality to the example app but am still working through the bureaucracy of shipping something I can test with end to end.

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java (hunk 5 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java:250]: Yes that is my understanding and what happened with "countryCode" in my last pr that was tested end to end.

File: packages/in_app_purchase/in_app_purchase_android/CHANGELOG.md (hunk 1 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/CHANGELOG.md]: Fixed, but I do think that it is weird that0.3.0+19 is less than 0.3.1

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/alternative_billing_only_reporting_details_wrapper.dart (hunk 6 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/alternative_billing_only_reporting_details_wrapper.dart:17]: fixed!

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart:1]: https://github.com/flutter/packages/pull/6056/files#diff-0c6c99864ca16044f0ab206454d1a8576e2fc12b3e8bf584af0c3b747376ad82

**Comment by stuartmorgan** [comment]
Is dependency resolution affected by Java version? I.e., could `androidx.test:runner` 1.5.2 need a version of Java that the CI doesn't have?

It looks like this is the first attempt to rev one of our packages past 1.4.0.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart [hunk 8]
```diff
@@ -157,4 +157,39 @@ class InAppPurchaseAndroidPlatformAddition
...
+  }
+
+  /// Disconnects, Sets AlternativeBillingOnly to true, and recoonects to
+  /// the [BillingClient].
+  ///
+  /// [alternativeBillingOnlyState] true will enable alternative billing only.
+  ///
+  /// Play apis have requirements for when this method can be called.
+  /// See: https://developer.android.com/google/play/billing/alternative/alternative-billing-without-user-choice-in-app
+  Future<void> setAlternativeBillingOnlyState(
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart:190]: This would also be a strong candidate for a more generic method with a mode enum. (I would say we should either do both this and BillingClientManager, or neither, for consistency.)

**Review by stuartmorgan — APPROVED** [review]
LGTM with some optional API feedback.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart [hunk 9]
```diff
@@ -119,6 +124,15 @@ class BillingClientManager {
     _purchasesUpdatedController.close();
   }
 
+  /// Ends connection to [BillingClient] and reconnects with [alternativeBillingState].
+  Future<void> reconnectWithAlternativeBillingOnlyState(
```
[the remark this answers is not in the document]
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart:128]: Yeah I am now looking at the api https://developer.android.com/reference/com/android/billingclient/api/BillingClient.Builder#enableUserChoiceBilling(com.android.billingclient.api.UserChoiceBillingListener)
They have this in a builder where you can set both user choice and alternative billing only which seems odd. 

I will create the enum and redo that section and its tests. Good idea

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart (hunk 8 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart:190]: yep will make an enum and modify it here and and in billing client manager. Then I will probably send down booleans with a documented priority to the java code.

**Comment by reidbaker** [comment]
Will request rereview after enum work is done (hopefully monday)

**Comment by reidbaker** [comment]
Requesting re-review because I had to modify a decent bit of code to support billing mode as an enum.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 10]
```diff
@@ -434,6 +489,46 @@ enum BillingResponse {
   networkError,
 }
 
+/// Plugin concept to cover billing modes.
+///
+/// [playBillingOnly] (google play billing only).
+/// [alternativeBillingOnly] (app provided billing with reporting to play).
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart:495]: Leave a comment here saying the values must match the BillingChoiceMode enum on the Java side?

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java [hunk 11]
```diff
@@ -65,10 +66,32 @@ static final class MethodNames {
...
+
+    // Key for an int argument passed into startConnection
+    static final String HANDLE = "handle";
+    // Key for a boolean argument passed into startConnection.
+    static final String BILLING_CHOICE_MODE = "billingChoiceMode";
+
+    private MethodArgs() {}
+  }
+
+  static final class BillingChoiceMode {
```
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java:90]: Leave a comment here saying the values must match the BillingChoiceMode enum on the Dart side?

**Review by stuartmorgan — APPROVED** [review]
LGTM

File: packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java [hunk 12]
```diff
@@ -199,25 +264,152 @@ public void getBillingConfigSuccess() {
...
 
-    verify(result, times(1)).success(fromBillingConfig(billingResult, mockBillingConfig));
+    verify(result).error(contains("UNAVAILABLE"), contains("BillingClient"), any());
+  }
+
+  @Test
+  public void createAlternativeBillingOnlyReportingDetailsSuccess() {
+    mockStartConnection();
+    // ArgumentCaptor<GetBillingConfigParams> paramsCaptor =
+    //     ArgumentCaptor.forClass(GetBillingConfigParams.class);
```
Comment (camsim99) [inline packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java:300]: Remove?

File: packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java [hunk 13]
```diff
@@ -866,8 +1058,13 @@ public void isFutureSupported_false() {
   }
 
   private ArgumentCaptor<BillingClientStateListener> mockStartConnection() {
+    return mockStartConnection(BillingChoiceMode.PLAY_BILLING_ONLY);
+  }
+
+  private ArgumentCaptor<BillingClientStateListener> mockStartConnection(int billingChoiceMode) {
```
Comment (camsim99) [inline packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java:1064]: Optional nit: Could you add a doc for this so it's easier to interpret why this may need to be called in some tests?

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart [hunk 14]
```diff
@@ -119,6 +121,19 @@ class BillingClientManager {
...
 
+  /// Ends connection to [BillingClient] and reconnects with [billingChoiceMode].
+  ///
+  /// Callers need to check if [BillingChoiceMode.alternativeBillingOnly] is
+  /// available by calling [BillingClientWrapper.isAlternativeBillingOnlyAvailable]
+  /// first.
+  Future<void> reconnectWithBillingChoiceMode(
+      BillingChoiceMode billingChoiceMode) async {
+    _billingChoiceMode = billingChoiceMode;
+    // Ends connection and triggers OnBillingServiceDisconnected, which causes reconnect;
```
Comment (camsim99) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart:132]: Nit: ```suggestion
    // Ends connection and triggers OnBillingServiceDisconnected, which causes reconnect:
```

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart [hunk 15]
```diff
@@ -157,4 +157,39 @@ class InAppPurchaseAndroidPlatformAddition
...
+  /// See: https://developer.android.com/reference/com/android/billingclient/api/BillingClient#showAlternativeBillingOnlyInformationDialog(android.app.Activity,%20com.android.billingclient.api.AlternativeBillingOnlyInformationDialogListener)
+  Future<BillingResultWrapper>
+      showAlternativeBillingOnlyInformationDialog() async {
+    final BillingResultWrapper wrapper =
+        await _billingClientManager.runWithClient((BillingClient client) =>
+            client.showAlternativeBillingOnlyInformationDialog());
+    return wrapper;
+  }
+
+  /// Disconnects, Sets AlternativeBillingOnly to true, and recoonects to
```
Comment (camsim99) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart:183]:
```suggestion
  /// Disconnects, sets AlternativeBillingOnly to true, and reconnects to
```

File: packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_addition_test.dart [hunk 16]
```diff
@@ -83,6 +85,90 @@ void main() {
...
+      stubPlatform.addResponse(
+        name: startConnectionCall,
+        additionalStepBeforeReturn: (dynamic value) =>
+            arguments = value as Map<dynamic, dynamic>,
+      );
+      stubPlatform.addResponse(name: endConnectionCall);
+      await iapAndroidPlatformAddition
+          .setBillingChoice(BillingChoiceMode.alternativeBillingOnly);
+
+      /// Fake the disconnect that we would expect from a endConnectionCall.
```
Comment (camsim99) [inline packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_addition_test.dart:101]:
```suggestion
      // Fake the disconnect that we would expect from a endConnectionCall.
```

File: packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_addition_test.dart [hunk 17]
```diff
@@ -83,6 +85,90 @@ void main() {
...
+      stubPlatform.addResponse(
+        name: startConnectionCall,
+        additionalStepBeforeReturn: (dynamic value) =>
+            arguments = value as Map<dynamic, dynamic>,
+      );
+      stubPlatform.addResponse(name: endConnectionCall);
+      await iapAndroidPlatformAddition
+          .setBillingChoice(BillingChoiceMode.playBillingOnly);
+
+      /// Fake the disconnect that we would expect from a endConnectionCall.
```
Comment (camsim99) [inline packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_addition_test.dart:125]:
```suggestion
      // Fake the disconnect that we would expect from a endConnectionCall.
```

**Review by camsim99 — APPROVED** [review]

File: packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java (hunk 13 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java:1072]: Done

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart (hunk 14 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart]: Done

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart (hunk 15 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform_addition.dart]: 🤦 thanks, fixed

File: packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java (hunk 12 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/android/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java]: done

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java (hunk 11 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/MethodCallHandlerImpl.java:94]: Good idea, done

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart (hunk 10 above)
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_wrapper.dart:495]: Good idea done.

File: packages/in_app_purchase/in_app_purchase_android/CHANGELOG.md [hunk 18]
```diff
@@ -1,3 +1,7 @@
+## 0.3.1
+
+* Adds alternative-billing-only APIs to platform addition.
```
[the remark this answers is not in the document]
Comment (reidbaker) [reply packages/in_app_purchase/in_app_purchase_android/CHANGELOG.md]: Done

**Comment by fareesh** [comment]
Seems there is a deadline to switch from manual reporting to the API by March 13. Is there a chance this will be in a release prior to that? Else we'll have to use a patched version or implement our own and then switch over

**Comment by stuartmorgan** [comment]
This was released three weeks ago.
